### ISSUE TITLE
Autonomous Indoor Flight with Computer Vision (Channel Override?)

### ISSUE THREAD

**Issue by GitWyd** [dronekit/dronekit-python#685]
We read that we should consult you first to figure out if there is a better solution than Channel Overrides for our problem.

We are a research team working on autonomous UAV flight in GPS deprived environments.

In order to do that, we need to

1. Arm the drone without GPS (is the "GUIDED_NOGPS" mode the right one for this?)
2. Send receiver commands to the drone: throttle, roll, pitch, yaw (Preferably with a magnitude) or other commands that allow us to directly control the drone i.e. change direction, ascend/descend, forwards/backwards

Eventually we will want to program aggressive flight maneuvers, but initially we simply need to have basic controls.

Does DroneKit provide some of those functions already? Which functions should we consider? Which functions do we need to code ourselves?

We are happy to contribute to the DroneKit project, if this is necessary to achieve the above described functionality.

Our Setup:
Navio2 with PI3 on custom build 330 size quadcopter

**Comment by peterbarker** [comment]
On Fri, 25 Nov 2016, GitWyd wrote:

>  1. Arm the drone without GPS (is the "GUIDED_NOGPS" mode the right one for
>     this?)

Probably.

>  2. Send receiver commands to the drone: throttle, roll, pitch, yaw
>     (Preferably with a magnitude) or other commands that allow us to
>     directly control the drone i.e. change direction, ascend/descend,
>     forwards/backwards

You probably don't want to do this; you probably want to send guided 
velocity mavlink messages instead: 
https://github.com/mavlink/mavlink/blob/master/message_definitions/v1.0/common.xml#L2962

> Eventually we will want to program aggressive flight maneuvers, but
> initially we simply need to have basic controls.
> 
> Does DroneKit provide some of those functions already? Which functions
> should we consider? Which functions do we need to code ourselves?

People have used dronekit to push their vehicles around using the guided 
velocity commands; 3DR's Solo vehicle's smartshots, for example.

**Comment by squilter** [comment]
`Guided-NoGPS` only allows you to send angle commands. You can send `SET_ATTITUDE_TARGET` messages, and you can listen to `ATTITUDE` messages.

If you attach an optical flow module and a rangefinder, then you can use normal guided mode. This will allow you to send velocity commands or even position commands.

**Comment by markrjr** [comment]
I have been doing some testing and the `SET_ATTITUDE_TARGET` messages do not work, and there's no way to set `Guided-NoGPS`.

**Comment by dollop80** [comment]
@squilter 

> If you attach an optical flow module and a rangefinder, then you can use normal guided mode. This will allow you to send velocity commands or even position commands.

I've added the OF and lidar. It works indoors without GPS **only** when arming in Loiter mode. No way to use OF in Guided mode. I've tried AC_3.4.5 and AC_3.5RC3

**Comment by studroid** [comment]
@dollop80 
Refer to #712 that I wrote.
This is probably what you are looking for!

**Comment by prasanna12510** [comment]
is there any simple code like arm and takeoff that demonstrates guided_nogps flight mode using dronekit api

**Comment by hamishwillee** [comment]
@prasanna12510 Yes. See https://github.com/dronekit/dronekit-python/pull/712 - not merged, but some great work there.

**Comment by borhanreo** [comment]
> @squilter
> 
> > If you attach an optical flow module and a rangefinder, then you can use normal guided mode. This will allow you to send velocity commands or even position commands.
> 
> I've added the OF and lidar. It works indoors without GPS **only** when arming in Loiter mode. No way to use OF in Guided mode. I've tried AC_3.4.5 and AC_3.5RC3

@dollop80 I have used maxbotix rangefinder sensor. So may i arm and takeoff in LOITER mode in indoor (GPS denied area).

**Comment by rajivrawat982** [comment]
I'm just looking to find out how flight controller deciding the actual altitude in "GUIDED_NOGPS" mode as while i was arming it without a props and using [vehicle.location.global_relative_frame.alt] to find out current altitude it changing continuously, it should not change as my drone is not moving only arming.
